### PR TITLE
feat: new rest api for command service

### PR DESCRIPTION
**Pull request by GregoryIvo**
This PR adds a new bundle which provides Command service support over REST. 

> **Note**: We are using the Conventional Commits convention for our pull request titles. Please take a look at the [PR title format document](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#submitting-the-changes) for the supported [types](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#type) and [scopes](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#scope).

Brief description of the PR. [e.g. Added `null` check on `object` to avoid `NullPointerException`]

**Related Issue:** n/a

**Description of the solution adopted:** n/a

**Screenshots:**  n/a

**Manual Tests**: n/a

**Any side note on the changes made:** n/a

### CODE REVIEW

**Comment by GregoryIvo** [comment]
This PR is still a WIP, and has not been tested.

File: kura/org.eclipse.kura.rest.command.provider/src/main/java/org/eclipse/kura/internal/rest/command/CommandRestService.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    private KuraMessage buildKuraMessage(RestCommandRequest restCommandPayload) {
+
+        KuraCommandRequestPayload kuraCommandRequestPayload = new KuraCommandRequestPayload(
+                restCommandPayload.getCommand());
+
+        kuraCommandRequestPayload.addMetric(PASSWORD_METRIC_NAME, new Password(restCommandPayload.getPassword()));
+        kuraCommandRequestPayload.setZipBytes(restCommandPayload.getZipBytes());
+
+        kuraCommandRequestPayload.setWorkingDir(restCommandPayload.getWorkingDirectory());
```
Comment (GregoryIvo) [inline kura/org.eclipse.kura.rest.command.provider/src/main/java/org/eclipse/kura/internal/rest/command/CommandRestService.java:94]: @nicolatimeus The variables after here are the 'optional' ones we discussed.

We do not need these extra vars below to conform to the Kapua web UI. They are also all optional and are not required to function. For the sake of completion I think we should keep them in the JSON, but then just ignore them client side.

Maybe can we put some feature flag (elevated permission?) here to disable them? 


What are your thoughts?

File: kura/org.eclipse.kura.rest.command.provider/src/main/java/org/eclipse/kura/rest/command/api/RestCommandRequest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ * SPDX-License-Identifier: EPL-2.0
+ * 
+ * Contributors:
+ *  Eurotech
+ *******************************************************************************/
+package org.eclipse.kura.rest.command.api;
+
+public class RestCommandRequest {
+
+    // Mandatory //
```
Comment (GregoryIvo) [inline kura/org.eclipse.kura.rest.command.provider/src/main/java/org/eclipse/kura/rest/command/api/RestCommandRequest.java]: I probably should use the Javadocs to mark these appropriately.

**Comment by GregoryIvo** [comment]
Progress:
image
Basic commands work!

**Comment by GregoryIvo** [comment]
docs can be found here: https://github.com/eclipse/kura/pull/4731

**Comment by GregoryIvo** [comment]
# Rest Command v1 API

#### Execute Command
- Method: POST
- API PATH: `/services/command/v1/command/`
##### Request Body
``` JSON
{
	//Command to be excuted on gateway
    "command":"printenv TextEnvVarName1",

    //Service Password for command Service
    "password":"s3curePassw0rd",

    //String base64 encoding of a zip file to transfer to gateway
    "zipBytes": "UEsDBAoACAAAAIyD1lYAA AAAAAAAAAAAAAAJACAAdGVzdGZpbGUxVVQNAAfprpRk6a6UZOmulGR1eAsAAQT1AQAABBQAAABQSwcIAAAAAAAAAAAAAAAAUEsBAgoDCgAIAAAAjIPWVgAAAAAAAAAAAAAAAAkAIAAAAAAAAAAAAKSBAAAAAHRlc3RmaWxlMVVUDQAH6a6UZOmulGTprpRkdXgLAAEE9QEAAAQUAAAAUEsFBgAAAAABAAEAVwAAAFcAAAAAAA==",

    //Command argument String array
    "arguments":["arg 1"],

    //Shell environment Pairs Map
    "environmentPairs": 
    {
        "TextEnvVarName1":"TextEnvVarValue1",
        "TextEnvVarName2":"TextEnvVarValue2"
    },
    //Working directory of command to be executed
    "workingDirectory":"/tmp",
    
    //Run command synchronously/asynchronously
    "isRunAsync":false
}
```

##### Responses
- 200 OK status

```JSON
{
    "stdout": "Command error output is displayed in this field",
    "stderr": "Command output is displayed in this field",
    "exitCode": 0,
    "isTimeOut": false
}
```

- 400 Bad Request (Malformed Client JSON)
- 404 Resource Not Found
- 500 Internal Server Error



!!! note

    Use the following command to retrieve the base64 representation of a zip file. `base64 -i <filename.zip>`

**Comment by GregoryIvo** [comment]
@nicolatimeus this PR is now ready for review!

File: kura/org.eclipse.kura.rest.command.provider/src/main/java/org/eclipse/kura/internal/rest/command/CommandRestService.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    private static final String KURA_PERMISSION_REST_CONFIGURATION_ROLE = "kura.permission.rest.command";
+
+    private static final String PASSWORD_METRIC_NAME = "command.password";
+    public static final String RESOURCE_COMMAND = "command";
+
+    private CommandCloudApp commandCloudApp;
+
+    public void setUserAdmin(UserAdmin userAdmin) {
+        userAdmin.createRole(KURA_PERMISSION_REST_CONFIGURATION_ROLE, Role.GROUP);
```
Comment (salvatore-coppola) [inline kura/org.eclipse.kura.rest.command.provider/src/main/java/org/eclipse/kura/internal/rest/command/CommandRestService.java:55]: What if the Configuration Role already exists ?

File: kura/org.eclipse.kura.rest.command.provider/META-INF/MANIFEST.MF [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ org.eclipse.kura.message;version="[1.0,2.0)",
+ org.eclipse.kura.cloudconnection.message;version="[1.0,2.0)",
+ org.eclipse.kura.request.handler.jaxrs;version="[1.0,2.0)",
+ org.eclipse.kura.request.handler.jaxrs.annotation;version="[1.0,2.0)",
+ org.osgi.framework;version="1.8.0",
+ org.osgi.service.cm;version="1.6.0",
+ org.osgi.service.component;version="1.3.0",
+ org.osgi.service.useradmin;version="1.1.0";resolution:=optional,
+ org.slf4j;version="1.7.25",
+ org.eclipse.kura.cloud.app.command;version="[1.3,2.0]"
```
Comment (salvatore-coppola) [inline kura/org.eclipse.kura.rest.command.provider/META-INF/MANIFEST.MF]: Check the version range here, it should be exclusive.

File: kura/test/org.eclipse.kura.rest.command.provider.test/META-INF/MANIFEST.MF [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ org.junit.rules;version="[4.12,5.0)",
+ org.junit.experimental.runners;version="[4.12,5.0)",
+ org.eclipse.kura.cloudconnection.message;version="[1.0,2.0)",
+ org.junit.runner;version="[4.12,5.0)",
+ org.junit.runners;version="[4.12,5.0)",
+ org.mockito;version="[4.0,5.0)",
+ org.mockito.invocation;version="[4.0,5.0)",
+ org.mockito.stubbing;version="[4.0,5.0)",
+ org.slf4j;version="1.6.4",
+ org.eclipse.kura.cloud.app.command;version="[1.3,2.0]"
```
Comment (salvatore-coppola) [inline kura/test/org.eclipse.kura.rest.command.provider.test/META-INF/MANIFEST.MF]: Check version Range here

File: kura/org.eclipse.kura.rest.command.provider/src/main/java/org/eclipse/kura/internal/rest/command/CommandRestService.java (hunk 3 above)
Comment (GregoryIvo) [reply kura/org.eclipse.kura.rest.command.provider/src/main/java/org/eclipse/kura/internal/rest/command/CommandRestService.java:55]: image
From OSGi 7 spec, I think this should be okay?

**Comment by GregoryIvo** [comment]
closed in favour of: https://github.com/eclipse/kura/pull/4743

**Pull request closed without merging**